parlai/core/torch_agent.py

```diff
@@ -1708,19 +1710,26 @@ def batch_act(self, observations):
         ``eval_step`` methods instead. The former is called when labels are
         present in the observations batch; otherwise, the latter is called.
         """
-        batch_size = len(observations)
+        if not isinstance(observations, Batch):
+            batch_size = len(observations)
+        else:
+            batch_size = observations.text_vec.size(0)
         # initialize a list of replies with this agent's id
         batch_reply = [
             Message({'id': self.getID(), 'episode_done': False})
             for _ in range(batch_size)
         ]
 
+        # create a batch from the vectors
+        if not isinstance(observations, Batch):
+            batch = self.batchify(observations)
+        else:
+            batch = observations
+
         # check if there are any labels available, if so we will train on them
+        observations = batch.observations
         self.is_training = any('labels' in obs for obs in observations)
 
-        # create a batch from the vectors
```

Comment:
Huh, why move this?

Reply:
I generalized to just access `observations` from the batch object now, since we need to use that in line 1731 (`self.is_training`)